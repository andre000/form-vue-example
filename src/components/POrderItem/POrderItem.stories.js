import { withKnobs, text, number } from '@storybook/addon-knobs'
import POrderItem from './POrderItem.vue'
import image from '@/components/PImageUpload/__mocks__/image'

export default {
  component: POrderItem,
  title: 'POrderItem',
  decorator: [withKnobs]
}

export const withInitialState = () => ({
  components: { POrderItem },
  props: {
    order: {
      default: {
        name: text('Name', 'Suco de Goiaba'),
        flavour: text('Flavour', 'Goiaba'),
        price: number('Price', '5'),
        description: text(
          'Description',
          'Um mero suco de goiaba. Tem sachet de açúcar no balcão...'
        ),
        image: {
          preview: image
        },
        type: 'drinks'
      }
    }
  },
  template: `<div style="position: relative"><p-order-item :order="order" /></div>`
})
