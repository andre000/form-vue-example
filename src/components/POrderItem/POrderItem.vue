<template>
  <li class="order-item">
    <div class="order-item__image">
      <img :src="orderImage" alt="" srcset="" />
    </div>
    <div class="order-item__header">
      <div class="order-item__name">"{{ order.name }}"</div>
      <div class="order-item__price">R$ {{ order.price | money }}</div>
    </div>
    <div class="order-item__body">
      <div>
        <span class="order-item__subtitle">Sabor: </span> {{ order.flavour }}
      </div>
      <div>
        <span class="order-item__subtitle">Descrição: </span>
        {{ order.description }}
      </div>
    </div>
  </li>
</template>

<script>
export default {
  name: 'POrderItem',
  props: {
    order: {
      type: Object,
      required: true,
      validator: order =>
        Object.keys(order).every(o =>
          // eslint-disable-next-line prettier/prettier
          ['name', 'flavour', 'price', 'description', 'image', 'type'].includes(o)
        ) && Object.keys(order).length === 6
    }
  },

  computed: {
    orderImage() {
      return (
        this.order.image?.preview || require('@/assets/img/pasteis-img.png')
      )
    }
  },

  filters: {
    money(v) {
      return Number(v).toLocaleString('pt-br', {
        mode: 'currency',
        currency: 'BRL'
      })
    }
  }
}
</script>

<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';

.order-item {
  box-shadow: 0px 0px 30px $shadow-color;
  border-radius: 20px;
  background: #fff;
  margin-left: 120px;

  &:not(:first-child) {
    margin-top: 50px;
  }

  .order-item__header {
    position: relative;
    background: $primary;
    border-radius: 20px 20px 0px 0px;
    padding: 20px 40px 20px 110px;
    display: flex;
    justify-content: space-between;

    .order-item__name {
      color: $secondary;
      font: 700 Italic 30px/37px Roboto;
    }
    .order-item__price {
      color: #fff;
      font: 700 Italic 26px/32px Roboto;
    }
  }

  .order-item__body {
    position: relative;
    padding: 36px 0px 36px 110px;
    color: $primary-text;
    font: 400 24px/32px Roboto;

    .order-item__subtitle {
      font-weight: bold;
      font-style: italic;
      margin-right: 10px;
    }
  }

  .order-item__image {
    position: absolute;
    box-shadow: 0px 0px 30px $shadow-color;
    border-radius: 20px;
    z-index: 10;
    background: #fff;
    left: 0px;
    margin-top: 21px;

    img {
      border-radius: 20px;
      width: 180px;
      height: 180px;
    }
  }
}
</style>
