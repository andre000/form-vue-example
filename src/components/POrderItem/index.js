export { default as POrderItem } from './POrderItem'
