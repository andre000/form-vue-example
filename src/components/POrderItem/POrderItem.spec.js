import { shallowMount, createLocalVue } from '@vue/test-utils'

import POrderItem from './POrderItem'

const localVue = createLocalVue()

const mockedOrder = {
  name: 'Suco de Goiaba',
  flavour: 'Goiaba',
  price: 5,
  description: 'Um mero suco de goiaba. Tem sachet de açúcar no balcão...',
  image: { preview: 'image' },
  type: 'drinks'
}

describe('POrderItem component', () => {
  let wrapper
  beforeAll(() => {
    wrapper = shallowMount(POrderItem, {
      localVue,
      propsData: { order: mockedOrder }
    })
  })

  it('should match the snapshot', () =>
    expect(wrapper.html()).toMatchSnapshot())

  it('should have an image element with the thumbnail of the order', () => {
    const thumb = wrapper.find('.order-item__image img')
    expect(thumb.exists()).toBe(true)
    expect(thumb.element.src).toMatch(mockedOrder.image.preview)
  })

  it('should show the order name and price', () => {
    const nameEl = wrapper.find('.order-item__name')
    const priceEl = wrapper.find('.order-item__price')

    expect(nameEl.exists()).toBe(true)
    expect(priceEl.exists()).toBe(true)

    expect(nameEl.text()).toBe(mockedOrder.name)
    expect(priceEl.text()).toBe(`${mockedOrder.price}`)
  })

  it('should show the order flavour and price', () => {
    const bodyEl = wrapper.find('.order-item__body')

    expect(bodyEl.exists()).toBe(true)

    expect(bodyEl.text()).toMatch(mockedOrder.flavour)
    expect(bodyEl.text()).toMatch(mockedOrder.description)
  })
})
