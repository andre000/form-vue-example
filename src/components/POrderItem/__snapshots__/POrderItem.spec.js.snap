// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POrderItem component should match the snapshot 1`] = `
<li class="order-item">
  <div class="order-item__image"><img src="image" alt="" srcset=""></div>
  <div class="order-item__header">
    <div class="order-item__name">"Suco de Goiaba"</div>
    <div class="order-item__price">R$ 5</div>
  </div>
  <div class="order-item__body">
    <div><span class="order-item__subtitle">Sabor: </span> Goiaba
    </div>
    <div><span class="order-item__subtitle">Descrição: </span>
      Um mero suco de goiaba. Tem sachet de açúcar no balcão...
    </div>
  </div>
</li>
`;
